trace/backend.go

```diff
@@ -0,0 +1,231 @@
+package trace
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"net"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/stripe/veneur/protocol"
+	"github.com/stripe/veneur/ssf"
+)
+
+// DefaultBackoff defaults to 10 milliseconds of initial wait
+// time. Subsequent wait times will add this backoff to the time they
+// wait.
+const DefaultBackoff time.Duration = 20 * time.Millisecond
+
+// DefaultMaxBackoff defaults to 1 second. No reconnection attempt
+// wait interval will be longer than this.
+const DefaultMaxBackoff time.Duration = 1 * time.Second
+
+// DefaultConnectTimeout is to 10 seconts. Any attempt to (re)connect
+// to a veneur will take longer than this. If it would take longer,
+// the span is discarded.
+const DefaultConnectTimeout time.Duration = 10 * time.Second
+
+// BufferSize is the default size of the SSF buffer. It defaults to
+// enough bytes to accomodate the largest SSF span.
+const BufferSize int = int(protocol.MaxSSFPacketLength + protocol.SSFFrameLength)
+
+type backendParams struct {
+	addr           net.Addr
+	backoff        time.Duration
+	maxBackoff     time.Duration
+	connectTimeout time.Duration
+	bufferSize     uint
+	cap            uint
+}
+
+func (p *backendParams) params() *backendParams {
+	return p
+}
+
+// backend is a structure that can send an SSF span to a destination
+// over a persistent connection, handling reconnections. When
+// encountering connection errors, a backend will automatically attempt
+// to reconnect and blocks until reconnecting succeeds.
+//
+// Data loss / resiliency to failure
+//
+// If a backend encounters an error sending a span, it discards
+// the span and attempts to reconnect. This means that the
+// backend is resilient against "poison pill" spans, but it
+// also means that spans will get lost if there are connection
+// problems, such as veneurs getting restarted.
+type backend interface {
+	io.Closer
+
+	params() *backendParams
+	connection(net.Conn)
+
+	// sendSync synchronously sends a span to an upstream veneur. If
+	// it encounters a protocol error in sending, it loops
+	// forever, backing off by n*the backoff interval (until it
+	// reaches the maximal backoff interval) and tries to
+	// reconnect. If sendSync encounters any non-protocol errors
+	// (e.g. in serializing the SSF span), it returns them without
+	// reconnecting.
```

Comment:
Nit: Interface types can't guarantee anything about the underlying implementation, so the documentation should ideally be worded to reflect that (e.g. "it should/must return them without reconnecting").